lucene/core/src/java/org/apache/lucene/search/QueueSizeBasedExecutionControlPlane.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.search;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Future;
+import java.util.concurrent.FutureTask;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.ThreadPoolExecutor;
+
+/**
+ * Implementation of SliceExecutionControlPlane with queue backpressure based thread allocation
+ */
+public class QueueSizeBasedExecutionControlPlane<C> implements SliceExecutionControlPlane<C> {
+  private static final double LIMITING_FACTOR = 1.5;
+  private static final int NUMBER_OF_PROCESSORS = Runtime.getRuntime().availableProcessors();
+
+  private Executor executor;
+
+  public QueueSizeBasedExecutionControlPlane(Executor executor) {
+    this.executor = executor;
+  }
+
+  @Override
+  public List<Future<C>> invokeAll(Collection<FutureTask> tasks) {
+    boolean isThresholdCheckEnabled = true;
+
+    if (tasks == null) {
+      throw new IllegalArgumentException("Tasks is null");
+    }
+
+    if (executor == null) {
+      throw new IllegalArgumentException("Executor is null");
+    }
+
+    ThreadPoolExecutor threadPoolExecutor = null;
+    if ((executor instanceof ThreadPoolExecutor) == false) {
```

Comment:
This is a bit trappy, if you wrap a ThreadPoolExecutor using the Executor API, you'd get different behavior than if you provided the ThreadPoolExecutor directly? Maybe this class should take a ThreadPoolExecutor, and there'd be a different simpler implementation for regular implementations. We'd keep an IndexSearcher constructor that takes an executor and documentation would say that we'd use this more sophisticated approach if it implements ThreadPoolExecutor?

Reply:
Agreed. Reverted the Executor changes and added the abstraction while updating the docs for IndexSearcher